src/components/List/List.tsx

```diff
@@ -0,0 +1,199 @@
+import classNames from "classnames";
+import PropTypes from "prop-types";
+import React from "react";
+import type { HTMLAttributes, ReactNode } from "react";
+
+import type { Headings } from "../../types";
+
+export type ListItem =
+  | ReactNode
+  | ({
+      content: ReactNode;
+    } & HTMLAttributes<HTMLLIElement>);
+
+export type SteppedListItem = {
+  content: ReactNode;
+  title: ReactNode;
+  titleElement?: Headings;
+} & HTMLAttributes<HTMLLIElement>;
+
+export type BaseProps = {
+  className?: string;
+  items: ListItem[];
+};
+
+export type DefaultListProps = BaseProps & {
+  detailed?: false;
+  divided?: boolean;
+  inline?: false;
+  isDark?: false;
+  middot?: false;
+  split?: boolean;
+  stepped?: false;
+  stretch?: false;
+  ticked?: boolean;
+};
+
+export type BaseInlineProps = BaseProps & {
+  detailed?: false;
+  divided?: false;
+  isDark?: boolean;
+  split?: false;
+  stepped?: false;
+  ticked?: false;
+};
+
+export type InlineListProps = BaseInlineProps & {
+  inline: true;
+  middot?: boolean;
+  stretch?: boolean;
+};
+
+export type MiddotListProps = BaseInlineProps & {
+  inline?: boolean;
+  middot: true;
+  stretch?: false;
+};
+
+export type StretchListProps = BaseInlineProps & {
+  inline?: boolean;
+  middot?: false;
+  stretch: true;
+};
+
+export type SteppedListProps = Exclude<BaseProps, "items"> & {
+  detailed?: boolean;
+  divided?: false;
+  inline?: false;
+  isDark?: false;
+  items: SteppedListItem[];
+  middot?: false;
+  split?: false;
+  stepped: true;
+  stretch?: false;
+  ticked?: false;
+} & HTMLAttributes<HTMLOListElement>;
+
+export type Props =
+  | ((DefaultListProps | InlineListProps | MiddotListProps | StretchListProps) &
+      HTMLAttributes<HTMLUListElement>)
+  | SteppedListProps;
```

Comment:
This is wild. Defs needs a refactor at some point.

Reply:
Haha yep! Defo.